pkg/state/snapshot_generator.go

```diff
@@ -1032,43 +1045,75 @@ func (sg *snapshotGenerator) generateBalancesAtomicSnapshots(
-	addrWavesBalanceDiff := make(addressWavesBalanceDiff)
+func (sg *snapshotGenerator) addAssetBalanceDiffFromTxDiff(change balanceDiff, assetKey []byte, scheme proto.Scheme,
+	addrAssetBalanceDiff addressAssetBalanceDiff) error {
+	if change.balance == 0 {
+		return nil
+	}
+	assetBalanceK := &assetBalanceKey{}
+	if err := assetBalanceK.unmarshal(assetKey); err != nil {
+		return errors.Errorf("failed to unmarshal asset balance key: %v", err)
+	}
+	asset := assetBalanceK.asset
+	address, cnvrtErr := assetBalanceK.address.ToWavesAddress(scheme)
+	if cnvrtErr != nil {
+		return errors.Wrap(cnvrtErr, "failed to convert address id to waves address")
+	}
+	assetBalKey := assetBalanceDiffKey{address: address, asset: asset}
+	addrAssetBalanceDiff[assetBalKey] = change.balance
+	// for compatibility with the legacy state hashes
+	sg.stor.balances.addAssetBalanceChangeLegacySH(address, asset, change.balance)
+	return nil
+}
+
+func (sg *snapshotGenerator) addWavesBalanceDiffFromTxDiff(change balanceDiff, wavesKey []byte, scheme proto.Scheme,
+	addrWavesBalanceDiff addressWavesBalanceDiff) error {
+	if change.balance == 0 && change.leaseOut == 0 && change.leaseIn == 0 {
+		return nil
+	}
+	wavesBalanceK := &wavesBalanceKey{}
+	if err := wavesBalanceK.unmarshal(wavesKey); err != nil {
+		return errors.Errorf("failed to unmarshal waves balance key: %v", err)
+	}
+	address, cnvrtErr := wavesBalanceK.address.ToWavesAddress(scheme)
+	if cnvrtErr != nil {
+		return errors.Wrap(cnvrtErr, "failed to convert address id to waves address")
+	}
+	addrWavesBalanceDiff[address] = change
+	// for compatibility with the legacy state hashes
+	sg.stor.balances.addWavesBalanceChangeLegacySH(address, change)
+	return nil
+}
+
+func (sg *snapshotGenerator) balanceDiffFromTxDiff(balanceChanges []balanceChanges,
+	scheme proto.Scheme) (addressWavesBalanceDiff, addressAssetBalanceDiff, error) {
 	addrAssetBalanceDiff := make(addressAssetBalanceDiff)
-	for balanceKeyString, diffAmount := range diff {
-		// construct address from key
-		wavesBalanceKey := &wavesBalanceKey{}
-		err := wavesBalanceKey.unmarshal([]byte(balanceKeyString))
-		if err != nil {
-			// if the waves balance unmarshal failed, try to marshal into asset balance, and if it fails, then return the error
-			assetBalanceKey := &assetBalanceKey{}
-			mrshlErr := assetBalanceKey.unmarshal([]byte(balanceKeyString))
-			if mrshlErr != nil {
-				return nil, nil, errors.Wrap(mrshlErr, "failed to convert balance key to asset balance key")
+	addrWavesBalanceDiff := make(addressWavesBalanceDiff)
+	for _, balanceChange := range balanceChanges {
+		// TODO check this place
+		if len(balanceChange.balanceDiffs) > 1 {
+			return nil, nil, errors.Errorf("more than one balance diff for the same address in the same block")
+		}
+		if len(balanceChange.key) > wavesBalanceKeySize {
+			err := sg.addAssetBalanceDiffFromTxDiff(balanceChange.balanceDiffs[0], balanceChange.key,
```

Comment:
balanceDiffs struct contains proto.BlockID, it means that inside a block, there's only one element in this array.
It used to be a diff for multiple blocks, but not anymore. The array should be replaced with a variable, but it would take a bit extra time to refactor code.

Reply:
Okay, for now we can leave it without changes.
I've just submitted a refactoring issue  #1297